src/collection.js

```diff
@@ -633,12 +687,22 @@ export default class Collection {
       lastModified: options.lastModified,
       headers: options.headers
     })
+      .then(changes => this.applyHook("incoming-changes", {changes}))
       // Reflect these changes locally
-      .then(changes => this.importChanges(syncResultObject, changes))
+      .then(payload => this.importChanges(syncResultObject, payload.changes))
       // Handle conflicts, if any
       .then(result => this._handleConflicts(result, options.strategy));
   }
 
+  applyHook(hookName, payload) {
+    if (typeof this.hooks[hookName] == "undefined") {
+      return Promise.resolve(payload);
+    }
+    return waterfall(this.hooks[hookName].map(hook => {
+      return record => hook(payload, this);
```

Comment:
[not finished reading the whole patch, parking this thought for now]: could we avoid passing `this` as a `hook` dependency here?